### PR TITLE
Fix automatic termination issue in EmrOperator by ensuring `waiter_max_attempts` is set for deferrable triggers

### PR DESCRIPTION
**Pull request by beobest2**
When running EMR jobs that require a long duration, setting the `waiter_max_attempts` in EmrOperator to a high value does not prevent the "waiter error: max attempts" error from occurring after a certain period.

```
[2024-03-27 16:42:08,646] Trigger failed:
Traceback (most recent call last):

  File "/usr/local/airflow/.local/lib/python3.11/site-packages/airflow/jobs/triggerer_job_runner.py", line 529, in cleanup_finished_triggers
    result = details["task"].result()
             ^^^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/airflow/.local/lib/python3.11/site-packages/airflow/jobs/triggerer_job_runner.py", line 601, in run_trigger
    async for event in trigger.run():

  File "/usr/local/airflow/.local/lib/python3.11/site-packages/airflow/providers/amazon/aws/triggers/base.py", line 143, in run
    await async_wait(

  File "/usr/local/airflow/.local/lib/python3.11/site-packages/airflow/providers/amazon/aws/utils/waiter_with_logging.py", line 133, in async_wait
    raise AirflowException("Waiter error: max attempts reached")
```
---

An issue has been identified with the AWS provider code.
The `EmrOperator` that calls this trigger does not allow users to input `waiter_max_attempts`, resulting in the application of the default value of 600.

`airflow/providers/amazon/aws/triggers/emr.py`
```python
class EmrContainerTrigger(AwsBaseWaiterTrigger):
    """
    Poll for the status of EMR container until reaches terminal state.
    :param virtual_cluster_id: Reference Emr cluster id
    :param job_id:  job_id to check the state
    :param aws_conn_id: Reference to AWS connection id
    :param waiter_delay: polling period in seconds to check for the status
    """

    def __init__(
        self,
        virtual_cluster_id: str,
        job_id: str,
        aws_conn_id: str | None = "aws_default",
        poll_interval: int | None = None,  # deprecated
        waiter_delay: int = 30,
        waiter_max_attempts: int = 600,   # <- We can't modify this ..
    ):
```

In other words, if the `poll_interval` is set to 60 seconds, the job will terminate after 10 hours (1 minute * 600) regardless of the `waiter_max_attempts` value specified by the user. 

To address this, modifications were made in `EmrOperator` to ensure that parameters provided by the user are also passed to the `EmrContainerTrigger`.

### CODE REVIEW

**Review by matcha-1atte — APPROVED** [review]
:+1:

**Review by matcha-1atte — APPROVED** [review]
Good.

**Review by vincbeck — APPROVED** [review]

**Comment by eladkal** [comment]
can you add unit tests to cover this change?

**Comment by beobest2** [comment]
@eladkal I have also added unit tests. Please review

File: tests/providers/amazon/aws/operators/test_emr_containers.py [hunk 1]
```diff
@@ -144,6 +144,20 @@ def test_operator_defer(self, mock_submit_job, mock_check_query_status):
             exc.value.trigger, EmrContainerTrigger
         ), f"{exc.value.trigger} is not a EmrContainerTrigger"
 
+    @mock.patch.object(EmrContainerHook, "submit_job")
+    @mock.patch.object(
+        EmrContainerHook, "check_query_status", return_value=EmrContainerHook.INTERMEDIATE_STATES[0]
+    )
+    def test_operator_defer_with_timeout(self, mock_submit_job, mock_check_query_status):
+        self.emr_container.deferrable = True
+        self.emr_container.max_polling_attempts = 1000
```
Comment (o-nikolas) [inline tests/providers/amazon/aws/operators/test_emr_containers.py:153]: 1000? How long does this unittest take? That seems excessive.

Comment (beobest2) [reply tests/providers/amazon/aws/operators/test_emr_containers.py:153]: @o-nikolas  The purpose of this PR was to address the issue where setting this value high would not wait beyond the default of 600. By setting it to 1000, the modification is proven to be implemented. The `poll_interval` is set to 0 for testing, so it does not increase the time it takes to run the tests.

Comment (o-nikolas) [reply tests/providers/amazon/aws/operators/test_emr_containers.py:153]: Ahh, the poll_interval is set to 0 in the setup_method, missed that. Looks good :+1:

**Review by o-nikolas — APPROVED** [review]

**Comment by o-nikolas** [comment]
Tests need to pass after your rebase from main. Otherwise looks good to merge

**Comment by o-nikolas** [comment]
Running workflows again after recent commits

**Comment by beobest2** [comment]
@o-nikolas  Previously, I was unable to properly test in my development environment because different dependency errors occurred. After solving the problem and running tests, I found that when deferrable is set to True, the `call_count` is not recorded as many times as the actual number of attempts we specified. Do you have an idea how can we verify the `call_count` in a unit test when using deferrable as it is now?

**Comment by o-nikolas** [comment]
> @o-nikolas Previously, I was unable to properly test in my development environment because different dependency errors occurred. After solving the problem and running tests, I found that when deferrable is set to True, the `call_count` is not recorded as many times as the actual number of attempts we specified. Do you have an idea how can we verify the `call_count` in a unit test when using deferrable as it is now?

CC @syedahsn

File: tests/providers/amazon/aws/sensors/test_emr_containers.py [hunk 2]
```diff
@@ -84,3 +84,15 @@ def test_sensor_defer(self, mock_poke):
...
         ), f"{e.value.trigger} is not a EmrContainerTrigger"
+
+    @mock.patch.object(
+        EmrContainerHook, "check_query_status", return_value=EmrContainerHook.INTERMEDIATE_STATES[0]
+    )
+    def test_sensor_defer_with_timeout(self, mock_check_query_status):
+        self.sensor.deferrable = True
+        self.sensor.max_polling_attempts = 1000
+
+        with pytest.raises(TaskDeferred):
```
Comment (syedahsn) [inline tests/providers/amazon/aws/sensors/test_emr_containers.py]: The problem with this test is that `check_query_status` is never called, because the sensor calls `self.defer` which raises the `TaskDeferred` Exception (which you are checking for), which ends the execution (i.e. the trigger code is never run).

What I would do here is just make sure that the Trigger is being called with the correct params (i.e. the supplied `max_polling_attempts`). The fact that the waiter calls the endpoint the correct number of times is technically the Trigger's responsibility. Because you are not overriding the `run` method of the `AwsBaseTrigger`, you should be fine with just ensuring that the correct parameters are being passed to the Trigger. The `run` method for the `AwsBaseTrigger` is tested in one [location](https://github.com/apache/airflow/blob/main/tests/providers/amazon/aws/triggers/test_base.py#L116). 
The same is true for the operator test above.

File: airflow/providers/amazon/aws/operators/emr.py [hunk 3]
```diff
@@ -617,6 +617,14 @@ def execute(self, context: Context) -> str | None:
                     job_id=self.job_id,
                     aws_conn_id=self.aws_conn_id,
                     waiter_delay=self.poll_interval,
+                    waiter_max_attempts=self.max_polling_attempts,
+                )
+                if self.max_polling_attempts
```
Comment (syedahsn) [inline airflow/providers/amazon/aws/operators/emr.py:622]: Wouldn't this be a better solution?
```
waiter_max_attempts=self.max_polling_attempts if self.max_polling_attempts else 600
```
You can write a test where you provide a `max_polling_attempts`, and one where you don't, and assert that the correct value is used.

**Review by syedahsn — COMMENTED** [review]
Left a couple of comments. Thanks for looking into this!

File: airflow/providers/amazon/aws/operators/emr.py (hunk 3 above)
Comment (beobest2) [reply airflow/providers/amazon/aws/operators/emr.py:622]: @syedahsn  I felt that using hardcoded values (such as 600) in the code can be problematic for maintenance. Especially when such a value is referenced in multiple places, if a change is necessary, finding and modifying all the reference points increases the likelihood of making errors. How do you think about this point?

File: tests/providers/amazon/aws/sensors/test_emr_containers.py (hunk 2 above)
Comment (beobest2) [reply tests/providers/amazon/aws/sensors/test_emr_containers.py]: I have modified the test code to only verify that the correct parameters are being passed to the trigger.

File: airflow/providers/amazon/aws/operators/emr.py (hunk 3 above)
Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/emr.py:622]: Hmm that's a good point. In that case, maybe you can edit the Trigger code to accept `None` (i.e. `waiter_max_attempts: int | None = 600`) as well, and do (I also included the `waiter_delay` param as well)


```
waiter_max_attempts = waiter_max_attempts or 600
waiter_delay = waiter_delay or 30
```
You would then be able to pass just pass `max_polling_attempts` as is, and the Trigger will handle the `None` case. What do you think about this approach?

File: tests/providers/amazon/aws/sensors/test_emr_containers.py [hunk 4]
```diff
@@ -84,3 +84,17 @@ def test_sensor_defer(self, mock_poke):
...
+        self.sensor.deferrable = True
+        mock_poke.return_value = False
+        self.sensor.max_retries = 1000
+
+        with pytest.raises(TaskDeferred) as e:
+            self.sensor.execute(context=None)
+
+        trigger = e.value.trigger
+        assert isinstance(trigger, EmrContainerTrigger), f"{trigger} is not a EmrContainerTrigger"
+        assert trigger.waiter_delay == self.sensor.poll_interval
```
Comment (syedahsn) [inline tests/providers/amazon/aws/sensors/test_emr_containers.py:99]: Nice! This looks great.

File: airflow/providers/amazon/aws/operators/emr.py (hunk 3 above)
Comment (beobest2) [reply airflow/providers/amazon/aws/operators/emr.py:622]: @syedahsn 
What you just mentioned was my initial approach. 
However, after analyzing the code, I found that modifying only the input type of [EmrStepSensorTrigger](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/triggers/emr.py#L213) results in a discrepancy with its parent class, violating the Liskov Substitution Principle (LSP). If we were to modify the parent class, [AwsBaseWaiterTrigger](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/triggers/base.py#L31), we would also need to change the type of `waiter_max_attempts` in the [async_await](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/utils/waiter_with_logging.py#L88) utility that accepts this input. This would require handling the exception when `waiter_max_attempts` is None. 
(https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/utils/waiter_with_logging.py#L119)
Additionally, this function is widely used across AWS operators, and we cannot predict what other side effects it might cause.

Considering these factors, I believe the current code modification, while not entirely clean, is the best approach.

Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/emr.py:622]: Alright I guess making those changes would go beyond the scope of this PR, so I agree. Great work with the deep dive!

**Review by syedahsn — APPROVED** [review]
LGTM
